utils/litemap/src/map.rs

```diff
@@ -0,0 +1,202 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/master/LICENSE ).
+
+use std::mem;
+use std::ops::{Index, IndexMut};
+
+/// A simple "flat" map based on a sorted vector
+///
+/// See the [module level documentation][super] for why one should use this.
+///
+/// The API is roughly similar to that of [`std::collections::HashMap`], though it
+/// requires `Ord` instead of `Hash`.
+#[derive(Clone, Debug)]
+pub struct LiteMap<K, V> {
+    values: Vec<(K, V)>,
+}
+
+impl<K, V> LiteMap<K, V> {
+    /// Construct a new [`LiteMap`]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// The number of elements in the [`LiteMap`]
+    pub fn len(&self) -> usize {
+        self.values.len()
+    }
+
+    /// Remove all elements from the [`LiteMap`]
+    pub fn clear(&mut self) {
+        self.values.clear()
+    }
+
+    /// Reserve capacity for `additional` more elements to be inserted into
+    /// the [`LiteMap`] to avoid frequent reallocations.
+    ///
+    /// See [`Vec::reserve()`] for more information.
+    pub fn reserve(&mut self, additional: usize) {
+        self.values.reserve(additional)
+    }
+}
+
+impl<K: Ord, V> LiteMap<K, V> {
+    /// Get the value associated with `key`, if it exists.
+    ///
+    /// ```rust
+    /// use litemap::LiteMap;
+    ///
+    /// let mut map = LiteMap::new();
+    /// map.insert(1, "one");
+    /// map.insert(2, "two");
+    /// assert_eq!(map.get(&1), Some(&"one"));
+    /// assert_eq!(map.get(&3), None);
+    /// ```
+    pub fn get(&self, key: &K) -> Option<&V> {
+        match self.values.binary_search_by(|k| k.0.cmp(&key)) {
+            Ok(found) => Some(&self.values[found].1),
+            Err(_) => None,
+        }
+    }
```

Comment:
I'm curious why we are not using `Borrow` here in the same way that `HashMap` and `BTreeMap` use `Borrow` for their key arguments:

```
    pub fn get<Q: ?Sized>(&self, key: &Q) -> Option<&V>
    where
        K: Borrow<Q>,
        Q: Ord,
```
https://doc.rust-lang.org/src/alloc/collections/btree/map.rs.html#520-530

Reply:
I thought of including it and wasn't sure if we'd need it so decided to not include it for v1. We can add it easily later. Since we're focusing on codesize it's better to avoid additional ways this can monomorphize, and it's not much work to ask people to "outline" their borrows.

I'm okay with adding it now if people think I should.

Reply:
I agree. Adding the functionality later would be completely backward compatible with the way its currently written.  I was just curious about the reasoning behind it.

My only argument for adding it would be In the interest of this crate being used by other projects as a general-purpose, lightweight map alternative. In that case, I think it would be good to be congruent with the way the standard library maps work.

Reply:
We should file a follow up issue to investigate if this has any binary size impact. My guess is that is would have minimal impact since the compiler should figure out borrows at compile time.

Reply:
Sounds good. I'll add it then.

Reply:
Done

Reply:
What is the issue number?